File: src/lib/golingvu/base_describe.go
package golingvu

// BaseDescribe extends BaseSetup for Describe-It pattern
type BaseDescribe struct {
	*BaseSetup
	Its []interface{}
}

// NewBaseDescribe creates a new BaseDescribe instance
func NewBaseDescribe(
	features []string,
	its []interface{},
	describeCB interface{},
	initialValues interface{},
) *BaseDescribe {
	// Map its to actions and checks
	// Since Its can mix mutations and assertions, we need to handle them differently
	baseSetup := NewBaseSetup(features, its, []interface{}{}, describeCB, initialValues)
	return &BaseDescribe{
		BaseSetup: baseSetup,
		Its:       its,
	}
}

// SetupThat implements BaseSetup's abstract method
func (bd *BaseDescribe) SetupThat(
	subject interface{},
	testResourceConfiguration ITTestResourceConfiguration,
	artifactory func(string, interface{}),
	setupCB interface{},
	initialValues interface{},
) (interface{}, error) {
	// This should be implemented by the adapter
	return subject, nil
}

// Describe is an alias for Setup for Describe-It pattern
func (bd *BaseDescribe) Describe(
	subject interface{},
	key string,
	testResourceConfiguration ITTestResourceConfiguration,
	tester func(interface{}) bool,
	artifactory func(string, interface{}),
	suiteNdx *int,
) (interface{}, error) {
	return bd.Setup(subject, key, testResourceConfiguration, tester, artifactory, suiteNdx)
}

File: src/lib/rubeno/lib/base_describe.rb
module Rubeno
  # BaseDescribe extends BaseSetup for Describe-It pattern (AAA).
  # Describe can be nested, and Its can mix mutations and assertions.
  class BaseDescribe < BaseSetup
    attr_accessor :its, :adapter
    
    def initialize(features, its, describe_cb, initial_values)
      # Since Its can mix mutations and assertions, we need to handle them differently
      super(features, its, [], describe_cb, initial_values)
      @its = its || []
      @adapter = nil
    end
    
    def setup_that(subject, test_resource_configuration, artifactory, setup_cb, initial_values)
      # This should be implemented by subclasses
      raise NotImplementedError, "Subclasses must implement setup_that"
    end
    
    def after_each(store, key, artifactory)
      if @adapter && @adapter.respond_to?(:after_each)
        @adapter.after_each(store, key, artifactory)
      else
        store
      end
    end
    
    # Alias for Describe-It pattern
    def describe(subject, key, test_resource_configuration, tester, artifactory = nil, suite_ndx = nil)
      setup(subject, key, test_resource_configuration, tester, artifactory, suite_ndx)
    end
  end
end
